Payment Channels (RIPD-1224)

seelabs opened the pull request:
This PR add unidirectional payment channels.

Payment channels permit off-ledger checkpoints of XRP payments flowing in a single direction. A channel sequesters the owner's XRP in its own ledger entry. The owner can authorize the recipient to claim up to a given balance by giving the receiver a signed message (off-ledger). The recipient can use this signed message to claim any unpaid balance while the channel remains open. The owner can top off the line as needed. If the channel has not paid out all its funds, the owner must wait out a delay to close the channel to give the recipient a chance to supply any claims. The recipient can close the channel at any time. Any transaction that touches the channel after the expiration time will close the channel. The total amount paid increases monotonically as newer claims are issued. When the channel is closed any remaining balance is returned to the owner. Channels are intended to permit intermittent off-ledger settlement of ILP trust lines as balances get substantial. For bidirectional channels, a payment channel can be used in each direction.

There are three new transactions (See the `PayChan.cpp` file for documentation on these Txs):
* PaymentChannelCreate
* PaymentChannelFund
* PaymentChannelClaim

There are three new RPC commands:
* account_channels queries the ledger for channels (much like account_lines)
* channel_authorize signs authorizations
* channel_verify verifying authorizations

JoelKatz commented on src/ripple/app/tx/impl/PayChan.cpp:
```diff
@@ -0,0 +1,470 @@
+            msg.add64 (authAmt.drops ());
+            if (!verify (pk, msg.slice (), *sig, /*canonical*/ true))
+                return temBAD_SIGNATURE;
+        }
```
You cannot check the signature in `doApply`. You must do it in `preclaim`. For the most obvious reason why it must be this way, consider what happens if we later have to retry the transaction against a different ledger. We don't want to have to check the signature again.

seelabs replied:
I moved the signature check into `preflight`. You suggested `preclaim`, but since the tx has the public key it can be checked without a ledger (the public key on the tx _is_ checked against the one stored in the ledger later.)

miguelportilla commented on src/ripple/rpc/handlers/AccountChannels.cpp:
```diff
@@ -0,0 +1,184 @@
+    }
+
+    unsigned int limit;
+    if (auto err = readLimitField(limit, RPC::Tuning::accountLines, context))
```
Use RPC::Tuning::accountChannels

miguelportilla commented on src/ripple/app/tests/PayChan_test.cpp:
```diff
@@ -0,0 +1,689 @@
+        expect (channelBalance (*env.current (), chan) == XRP (0));
+        expect (channelAmount (*env.current (), chan) == XRP (1000));
+
+        env (fund (alice, chan, XRP (1000)));
```
Consider testing a non XPR with fund (should be temBAD_AMOUNT?).

miguelportilla commented on src/ripple/app/tests/PayChan_test.cpp:
```diff
@@ -0,0 +1,689 @@
+        env.fund (XRP (10000), alice, bob);
+        auto const pk = alice.pk ();
+        auto const settleDelay = 100s;
+        env (create (alice, bob, XRP (1000), settleDelay, pk));
```
Consider testing a non XPR with fund (should be temBAD_AMOUNT?).

miguelportilla commented on src/ripple/app/tests/PayChan_test.cpp:
```diff
@@ -0,0 +1,689 @@
+            auto const reqBal = chanBal + delta;
+            auto const authAmt = reqBal + XRP (100);
+            assert (reqBal <= chanAmt);
+            env (claim (alice, chan, reqBal, authAmt));
```
Since the TransactionType is different from create and fund, consider testing non XPR with claim.

miguelportilla replied:
Attempt a no signature claim for more than authorized? I realize a signed claim for more than authorized is tested later on.

miguelportilla replied:
Also consider testing negative XRP values with the create, fund and claim transactions.

miguelportilla commented on src/ripple/app/tests/PayChan_test.cpp:
```diff
@@ -0,0 +1,689 @@
+            auto const delta = XRP (500);
+            auto const reqBal = chanBal + delta;
+            auto const authAmt = reqBal + XRP (100);
+            assert (reqBal <= chanAmt && reqBal <= authAmt);
```
'reqBal <= authAmt' ?

vinniefalco commented on src/ripple/app/tests/PayChan_test.cpp:
```diff
@@ -0,0 +1,752 @@
+    channelExpiration (ReadView const& view, uint256 const& chan)
+    {
+        Keylet const k (ltPAYCHAN, chan);
+        auto const slep = view.read (k);
```
Consider
```
auto const sle = view.read({ltPAYCHAN, id});
```

miguelportilla commented on src/ripple/app/tests/PayChan_test.cpp:
```diff
@@ -0,0 +1,689 @@
+            expect (channelBalance (*env.current (), chan) == chanBal);
+            expect (channelAmount (*env.current (), chan) == chanAmt);
+            auto const feeDrops = env.current ()->fees ().base;
+            expect (env.balance (bob) == preBob + -feeDrops);
```
Should be 'preBob - feeDrops'

miguelportilla commented on src/ripple/app/tests/PayChan_test.cpp:
```diff
@@ -0,0 +1,689 @@
+            expect (!channelExists (*env.current (), chan));
+            auto const feeDrops = env.current ()->fees ().base;
+            expect (env.balance (alice) == preAlice + chanAmt - chanBal);
+            expect (env.balance (bob) == preBob + - feeDrops);
```
Should be 'preBob - feeDrops'

miguelportilla commented on src/ripple/app/tests/PayChan_test.cpp:
```diff
@@ -0,0 +1,689 @@
+            chanBal = reqBal;
+        }
+        {
+            auto preBob = env.balance (bob);
```
Suggest comment: // claim with signature

miguelportilla commented on src/ripple/app/tests/PayChan_test.cpp:
```diff
@@ -0,0 +1,689 @@
+        expect (channelAmount (*env.current (), chan) == XRP (1000));
+
+        env (fund (alice, chan, XRP (1000)));
+        auto chanBal = channelBalance (*env.current (), chan);
```
Should probably check alice's balance here, just to verify all is peachy.

miguelportilla commented on src/ripple/app/tests/PayChan_test.cpp:
```diff
@@ -0,0 +1,689 @@
+        env (fund (alice, chan, XRP (1000)));
+        auto chanBal = channelBalance (*env.current (), chan);
+        auto chanAmt = channelAmount (*env.current (), chan);
+        expect (chanBal == XRP (0));
```
Should probably check alice's balance here, just to verify all is peachy.

miguelportilla commented on src/ripple/app/tests/PayChan_test.cpp:
```diff
@@ -0,0 +1,689 @@
+        auto chanAmt = channelAmount (*env.current (), chan);
+        expect (chanBal == XRP (0));
+        expect (chanAmt == XRP (2000));
+
```
Should probably check alice's balance here, just to verify all is peachy.

miguelportilla commented on src/ripple/app/tx/impl/PayChan.cpp:
```diff
@@ -0,0 +1,470 @@
+        DestinationTag (optional)
+            Required if the destination has lsfRequireDestTag set.
+        SourceTag (optional)
+
```
Description?

miguelportilla commented on src/ripple/app/tx/impl/PayChan.cpp:
```diff
@@ -0,0 +1,470 @@
+TER
+PayChanCreate::preflight (PreflightContext const& ctx)
+{
+    if (!ctx.rules.enabled (featurePayChan, ctx.app.config ().features))
```
We seem to be allowing negative values through to sfSettleDelay. Since its unsigned, we might end up with an erroneous value.

seelabs replied:
By the time we get to preflight, it's too late to do this check. I think we'd need a check when the json is converted to an STObject. If we want to do this it sounds like this should be a general check and not something specific to payment channels. Does anyone know if we already have these checks in place? Do we want to add them?

seelabs replied:
@miguelportilla There appears to be a bug in `range_check_cast`. I opened a separate PR to address the issue.

miguelportilla commented on src/ripple/app/tests/PayChan_test.cpp:
```diff
@@ -0,0 +1,689 @@
+        testDefaultAmount ();
+        testDisallowXRP ();
+        testDstTag ();
+        testMultiple ();
```
Add test for SrcTag?

miguelportilla commented on src/ripple/app/tests/PayChan_test.cpp:
```diff
@@ -0,0 +1,689 @@
+        testCloseDry ();
+        testDefaultAmount ();
+        testDisallowXRP ();
+        testDstTag ();
```
Add test for SrcTag?

seelabs replied:
There are situations where the dst tag is required, and we test for those. There are no analogous tests for src tag.

nbougalis commented on src/ripple/app/tests/PayChan_test.cpp:
```diff
@@ -0,0 +1,689 @@
+        STAmount const& authAmt)
+    {
+        Serializer msg;
+        std::uint8_t const hdr[4] = {'C', 'L', 'M', 0};
```
Should this be a `HashPrefix`? See: https://github.com/ripple/rippled/blob/develop/src/ripple/protocol/impl/HashPrefix.cpp

nbougalis commented on src/ripple/app/tests/PayChan_test.cpp:
```diff
@@ -0,0 +1,689 @@
+    channelBalance (ReadView const& view, uint256 const& chan)
+    {
+        Keylet const k (ltPAYCHAN, chan);
+        auto const slep = view.read (k);
```
`auto const slep = view.read ({ltPAYCHAN, chan});`

miguelportilla commented on src/ripple/app/tx/impl/PayChan.cpp:
```diff
@@ -0,0 +1,470 @@
+        return ret;
+
+    auto const bal = ctx.tx[~sfBalance];
+    auto const amt = ctx.tx[~sfAmount];
```
Suggest declaring only before use.

nbougalis commented on src/ripple/app/tx/impl/PayChan.cpp:
```diff
@@ -0,0 +1,470 @@
+
+    // Transfer amount back to owner, decrement owner count
+    auto const sle = view.peek (keylet::account (src));
+    if (!sle)
```
Under which condition is this possible?

seelabs replied:
It can't happen. I'll remove the check.

nbougalis commented on src/ripple/app/tx/impl/PayChan.cpp:
```diff
@@ -0,0 +1,470 @@
+        return temBAD_AMOUNT;
+
+    if (ctx.tx[sfAccount] == ctx.tx[sfDestination])
+        return temMALFORMED;
```
Use `temDST_IS_SRC` instead.

miguelportilla commented on src/ripple/app/tests/PayChan_test.cpp:
```diff
@@ -668,6 +668,68 @@ struct PayChan_test : public beast::unit_test::suite
+                    env.rpc ("account_channels", alice.human (), bob.human ());
+            expect (r[jss::result][jss::channels].size () == 2);
+            expect (chan1Str != chan2Str);
+            for(std::uint32_t i = 0; i<2; ++i)
```
I'd prefer the following. Especially since chas2Str can be absent and the loop would pass.
expect(r[jss::result][jss::channels][0u][jss::channel_id] == chan2Str ||
            r[jss::result][jss::channels][1u][jss::channel_id] == chan2Str);

miguelportilla commented on src/ripple/app/tests/PayChan_test.cpp:
```diff
@@ -0,0 +1,689 @@
+        auto const chan = channel (*env.current (), alice, bob);
+        expect (channelBalance (*env.current (), chan) == XRP (0));
+        expect (channelAmount (*env.current (), chan) == XRP (1000));
+
```
Should probably check alice's balance after create, to verify the amount was sequestered.

nbougalis commented on src/ripple/protocol/impl/TxFormats.cpp:
```diff
@@ -113,6 +113,26 @@ TxFormats::TxFormats ()
+            SOElement (sfSettleDelay,       SOE_REQUIRED) <<
+            SOElement (sfPublicKey,         SOE_REQUIRED) <<
+            SOElement (sfCancelAfter,       SOE_OPTIONAL) <<
+            SOElement (sfDestinationTag,    SOE_OPTIONAL);
```
I'd suggest allowing `sfInvoiceID` as an optional field as well: it's a `uint256` which is _almost_ orthogonal to the destination tag and affords a lot more space for tracking multiple concurrent flows and associating a particular channel with a unique, user-selectable identifier.

seelabs replied:
@JoelKatz What do you think about this? Should I add the invoice field?

nbougalis commented on src/ripple/app/tx/impl/PayChan.cpp:
```diff
@@ -0,0 +1,470 @@
+
+        if (balance < reserve + ctx_.tx[sfAmount])
+            return tecUNFUNDED;
+    }
```
I believe that this block, as well as the next (so lines 176-201) can be handled in `preclaim`. Is there a reason why we do this in `doApply`?

nbougalis commented on src/ripple/app/tx/impl/PayChan.cpp:
```diff
@@ -0,0 +1,470 @@
+            ctx_.view ().info ().parentCloseTime.time_since_epoch ().count ();
+        if ((cancelAfter && closeTime >= *cancelAfter) ||
+            (expiration && closeTime >= *expiration))
+            return closeChannel (
```
I have concerns: a "fund" operation on an expired channel will return `tesSUCCESS` from here; this makes deciding what happened a nightmare. This is a problem. I suggest we add a new `tecEXPIRED` code and do something like the following:

```
auto ret = closeChannel (
    slep, ctx_.view (), k.key, ctx_.app.journal ("View"));

if (isTesSuccess(ret))
    ret = tecEXPIRED;

return ret;
```

nbougalis replied:
Addendum: `tesSUCCESS` is the best we can do now because `closeChannel` will transfer any unclaimed funds back to the source,and a `tec` code can't really perform transactions.

The documentation needs to explicitly explain that "fund" really means "fund if not expired, do nothing otherwise" and that a `tesSUCCESS` may mean "funding succeeded!" or "channel expired."

nbougalis commented on src/ripple/app/tx/impl/PayChan.cpp:
```diff
@@ -0,0 +1,470 @@
+        auto minExpiration =
+            ctx_.view ().info ().parentCloseTime.time_since_epoch ().count () +
+                (*slep)[sfSettleDelay];
+        if (expiration && *expiration < minExpiration)
```
Little known-fact: you can skip the `*` here: `boost::optional` transparently overloads operators. See: http://www.boost.org/doc/libs/1_61_0/libs/optional/doc/html/boost_optional/tutorial/relational_operators.html

I'm not suggesting you change anything; just pointing out it's possible. Caveat emptor: I don't know if `std::optional` will have the same behavior.

seelabs replied:
Yeah, I knew about that (I know because Scott Schurr used it in some code I reviewed). I prefer the non-magic version.

nbougalis commented on src/ripple/app/tx/impl/PayChan.cpp:
```diff
@@ -0,0 +1,470 @@
+            ctx_.view ().info ().parentCloseTime.time_since_epoch ().count ();
+        if ((cancelAfter && closeTime >= *cancelAfter) ||
+            (curExpiration && closeTime >= *curExpiration))
+            return closeChannel (
```
Same concern as the `fund` case. Returning `tesSUCCESS` if I try to claim on an expired channels complicates things. I'd prefer `tecEXPIRED`.

JoelKatz replied:
We have no choice but to return `tesSUCCESS` if we do anything other than claim a fee. That's the nature of the design at this point.

nbougalis replied:
Yep. I'm convinced.

nbougalis commented on src/ripple/app/tx/impl/PayChan.cpp:
```diff
@@ -0,0 +1,470 @@
+            if (ctx_.tx[sfPublicKey] != pk)
+                return temBAD_SIGNER;
+            Serializer msg;
+            std::uint8_t const hdr[4] = {'C', 'L', 'M', 0};
```
We duplicate this bit of code in the test, as part of `signClaimAuth`. I would suggest having a function that does this once:

```
void serializePayChanMessage (
    Serializer& msg,
    uint256 const& key,
    XRPAmount const& amount)
{
    std::uint8_t const hdr[4] = {'C', 'L', 'M', 0};
    msg.addRaw (hdr, 4);
    msg.add256 (key);
    msg.add64 (amount.drops());
}
```

If you don't want to pass the `Serializer` and are ok with a memory allocation and a copy, you can do:

```
Buffer serializePayChanMessage (
    uint256 const& key,
    XRPAmount const& amount)
{
    std::uint8_t const hdr[4] = {'C', 'L', 'M', 0};

    Serializer msg;
    msg.addRaw (hdr, 4);
    msg.add256 (key);
    msg.add64 (amount.drops());

    return { msg.slice() };
}
```

nbougalis commented on src/ripple/app/tx/impl/PayChan.cpp:
```diff
@@ -0,0 +1,470 @@
+            msg.addRaw (hdr, 4);
+            msg.add256 (k.key);
+            msg.add64 (authAmt.drops ());
+            if (!verify (pk, msg.slice (), *sig, /*canonical*/ true))
```
The `true` isn't strictly necessary - it's defaulted and unlikely to ever change, so you could skip it. On the other hand, we want to enforce canonical signatures, so "reinforcing" that by explicitly demanding canonical verification isn't a bad idea.

I'm not feeling strongly about this either way.

nbougalis commented on src/ripple/app/tx/impl/PayChan.cpp:
```diff
@@ -0,0 +1,477 @@
+    if ((flags & tfClose) && (flags & tfRenew))
+        return temMALFORMED;
+
+    if (auto const& sig = ctx.tx[~sfSignature])
```
Why `&` here?

seelabs replied:
No reason really. I changed it to a value (<pedantic> The ref doesn't _hurt_ us here, but given the two equivalent ways I agree a value type here is expected).

seelabs commented:
@nbougalis @miguelportilla I pushed two new commits that should address all the review feedback except two:

1) Mickey's suggestion to check for negative settle times
2) Nik's suggestion to add a sfInvoiceID field

seelabs commented:
Rebased onto 0.33.0-b1

seelabs commented:
Rebased onto 0.33.0-b2

seelabs commented:
@nbougalis @miguelportilla Added a sfInvoiceID field (as suggested by Nik - David thought it was a good idea as well). I believe that addresses all the feedback.

seelabs commented:
Pushed a commit to add an InvoiceID to the claim tx. If the channel is created with a InvoiceID, the claim tx must must have the matching InvoiceID. If it does not `temBAD_INVOICE` is returned.

In addition to a code review, please think about if the interface feels right. Let's get the interface right or add invoices later if we can't get it right now.

miguelportilla commented on src/ripple/app/tx/impl/PayChan.cpp:
```diff
@@ -358,7 +374,7 @@ PayChanClaim::preflight (PreflightContext const& ctx)
         auto const reqBalance = bal->xrp ();
         auto const authAmt = [&] {
-            if (auto const& amt = ctx.tx[~sfAmount])
+            if (auto const amt = ctx.tx[~sfAmount])
```
amt was already declared on 355 and you're auto capturing by ref.

miguelportilla commented on src/ripple/app/tests/PayChan_test.cpp:
```diff
@@ -106,7 +106,8 @@ struct PayChan_test : public beast::unit_test::suite
         boost::optional<NetClock::time_point> cancelAfter = boost::none,
-        boost::optional<std::uint32_t> dstTag = boost::none)
+        boost::optional<std::uint32_t> dstTag = boost::none,
+        boost::optional<uint256> invoiceID = boost::none)
```
Super nit pick: These can both be const refs.

seelabs replied:
Already fixed in a later commit.

nbougalis commented on src/ripple/app/tx/impl/PayChan.cpp:
```diff
@@ -0,0 +1,512 @@
+        // present, check it
+
+        auto const reqBalance = bal->xrp ();
+        auto const authAmt = [&] {
```
Why not:

    auto const authAmt = (amt ? amt->xrp() : reqBalance);

miguelportilla commented on src/ripple/app/tests/PayChan_test.cpp:
```diff
@@ -779,6 +782,64 @@ struct PayChan_test : public beast::unit_test::suite
+        }
+        {
+            Blob raw(256/8);
+            std::iota(raw.begin(), raw.end(), 0);
```
#include <numeric>

seelabs commented:
@nbougalis @miguelportilla I pushed a commit that removes the InvoiceID additions and leaves some minor cleanups.

miguelportilla commented:
:+1:

seelabs commented:
Squashed and marking as passed (I spoke with @nbougalis ; He gave a thumbs-up in a DM).
